presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java

```diff
@@ -1139,7 +1139,7 @@ public PhysicalOperation visitProject(ProjectNode node, LocalExecutionPlanContex
                 sourceNode = filterNode.getSource();
                 filterExpression = Optional.of(filterNode.getPredicate());
                 Expression temp = filterNode.getPredicateWithoutTupleDomain();
-                if (!temp.equals(TRUE_LITERAL)) {
+                if (temp != null && !temp.equals(TRUE_LITERAL)) {
```

Comment:
@yingsu00 Ying, could you explain why `temp` can be null now?

Reply:
@mbasmanova Hi Masha, the Expression temp was added by Orri to get the predicate without TupleDomain(not expressible by TupleDomain). For queries like show session, there is no such predicate.

Reply:
@yingsu00 Ying, thanks for clarifying. I took a closer look and I couldn't see any caller that would create an instance of FilterNode with `null` value for the `predicateWithoutTupleDomain` field. I also tried to reproduce this error by running `TpchQueryRunner` and `Presto` CLI from the IDEA and couldn't.

How did you reproduce this issue?

I think a better fix is to add checks to `FilterNode` constructor to ensure that `predicateWithoutTupleDomain` is never null, e.g. `requireNonNull(predicateWithoutTupleDomain, "predicateWithoutTupleDomain is null");` and update the callers which handle `nulls`, e.g. `com.facebook.presto.sql.planner.optimizations.UnaliasSymbolReferences.Rewriter#visitFilter`:

```
        @Override
        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
        {
            PlanNode source = context.rewrite(node.getSource());
            Expression predicate2 = node.getPredicateWithoutTupleDomain();
            return new FilterNode(node.getId(), source, canonicalize(node.getPredicate()), predicate2 != null ? canonicalize(predicate2) : null);
        }
```

should be

```
        @Override
        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
        {
            PlanNode source = context.rewrite(node.getSource());
            return new FilterNode(node.getId(), source, canonicalize(node.getPredicate()), canonicalize(node.getPredicateWithoutTupleDomain()));
        }
```

Here is the output of my test:

```
presto> show session
show session
     -> ;
;
                     Name                     |         Value         |        Default        |  Type   |                                                      Description                                                       
----------------------------------------------+-----------------------+-----------------------+---------+------------------------------------------------------------------------------------------------------------------------
 aggregation_operator_unspill_memory_limit    | 4MB                   | 4MB                   | varchar | Experimental: How much memory can should be allocated per aggragation operator in unspilling process                   
 aria                                         | true                  | true                  | boolean | Enable Aria Presto! scan operator                                                                                      
 aria_flags                                   | 0                     | 0                     | integer | Enable various Aria Presto! experiments                                                                                
 aria_reorder                                 | true                  | true                  | boolean | Enable Aria Presto! reorder of filters                                                                                 
 aria_reuse_pages                             | true                  | true                  | boolean | Enable Aria Presto! reuse of Pages                                                                                     
 colocated_join                               | false                 | false                 | boolean | Experimental: Use a colocated join when possible                                                                       
 concurrent_lifespans_per_task                | 0                     | 0                     | integer | Experimental: Run a fixed number of groups concurrently for eligible JOINs                                             
 default_filter_factor_enabled                | false                 | false                 | boolean | use a default filter factor for unknown filters in a filter node                                                       
 dictionary_aggregation                       | false                 | false                 | boolean | Enable optimization for aggregations on dictionaries                                                                   
 distributed_index_join                       | true                  | true                  | boolean | Distribute index joins on join keys instead of executing inline                                                        
 distributed_join                             |                       |                       | boolean | (DEPRECATED) Use a distributed join instead of a broadcast join. If this is set, join_distribution_type is ignored.    
 distributed_sort                             | true                  | true                  | boolean | Parallelize sort across multiple nodes                                                                                 
 dynamic_schedule_for_grouped_execution       | false                 | false                 | boolean | Experimental: Use dynamic schedule for grouped execution when possible                                                 
 enable_intermediate_aggregations             | false                 | false                 | boolean | Enable the use of intermediate aggregations                                                                            
 enable_stats_calculator                      | true                  | true                  | boolean | Experimental: Enable statistics calculator                                                                             
 exchange_compression                         | false                 | false                 | boolean | Enable compression in exchanges                                                                                        
 execution_policy                             | all-at-once           | all-at-once           | varchar | Policy used for scheduling query tasks                                                                                 
 fast_inequality_joins                        | true                  | true                  | boolean | Use faster handling of inequality join if it is possible                                                               
 filter_and_project_min_output_page_row_count | 256                   | 256                   | integer | Experimental: Minimum output page row count for filter and project operators                                           
 filter_and_project_min_output_page_size      | 500kB                 | 500kB                 | varchar | Experimental: Minimum output page size for filter and project operators                                                
 grouped_execution_for_aggregation            | false                 | false                 | boolean | Use grouped execution for aggregation when possible                                                                    
 hash_partition_count                         | 100                   | 100                   | integer | Number of partitions for distributed joins and aggregations                                                            
 ignore_stats_calculator_failures             | true                  | true                  | boolean | Ignore statistics calculator failures                                                                                  
 initial_splits_per_node                      | 4                     | 4                     | integer | The number of splits each node will run per task, initially                                                            
 iterative_optimizer_enabled                  | true                  | true                  | boolean | Experimental: enable iterative optimizer                                                                               
 iterative_optimizer_timeout                  | 3.00m                 | 3.00m                 | varchar | Timeout for plan optimization in iterative optimizer                                                                   
 join_distribution_type                       | PARTITIONED           | PARTITIONED           | varchar | The join method to use. Options are BROADCAST,PARTITIONED,AUTOMATIC                                                    
 join_reordering_strategy                     | ELIMINATE_CROSS_JOINS | ELIMINATE_CROSS_JOINS | varchar | The join reordering strategy to use. Options are NONE,ELIMINATE_CROSS_JOINS,AUTOMATIC                                  
 legacy_row_field_ordinal_access              | false                 | false                 | boolean | Allow accessing anonymous row field with .field0, .field1, ...                                                         
 legacy_unnest                                | false                 | false                 | boolean | Using legacy unnest semantic, where unnest(array(row)) will create one column of type row                              
 max_drivers_per_task                         |                       |                       | integer | Maximum number of drivers per task                                                                                     
 max_reordered_joins                          | 9                     | 9                     | bigint  | The maximum number of joins to reorder as one group in cost-based join reordering                                      
 optimize_hash_generation                     | true                  | true                  | boolean | Compute hash codes for distribution, joins, and aggregations early in query plan                                       
 optimize_metadata_queries                    | false                 | false                 | boolean | Enable optimization for metadata queries                                                                               
 optimize_mixed_distinct_aggregations         | false                 | false                 | boolean | Optimize mixed non-distinct and distinct aggregations                                                                  
 optimize_top_n_row_number                    | true                  | true                  | boolean | Use top N row number optimization                                                                                      
 parse_decimal_literals_as_double             | false                 | false                 | boolean | Parse decimal literals as DOUBLE instead of DECIMAL                                                                    
 plan_with_table_node_partitioning            | true                  | true                  | boolean | Experimental: Adapt plan to pre-partitioned tables                                                                     
 prefer_partial_aggregation                   | true                  | true                  | boolean | Prefer splitting aggregations into partial and final stages                                                            
 prefer_streaming_operators                   | false                 | false                 | boolean | Prefer source table layouts that produce streaming operators                                                           
 push_aggregation_through_join                | true                  | true                  | boolean | Allow pushing aggregations below joins                                                                                 
 push_partial_aggregation_through_join        | false                 | false                 | boolean | Push partial aggregations below joins                                                                                  
 push_table_write_through_union               | true                  | true                  | boolean | Parallelize writes when using UNION ALL in queries that write data                                                     
 query_max_cpu_time                           | 1000000000.00d        | 1000000000.00d        | varchar | Maximum CPU time of a query                                                                                            
 query_max_execution_time                     | 100.00d               | 100.00d               | varchar | Maximum execution time of a query                                                                                      
 query_max_run_time                           | 100.00d               | 100.00d               | varchar | Maximum run time of a query (includes the queueing time)                                                               
 query_priority                               | 1                     | 1                     | integer | The priority of queries. Larger numbers are higher priority                                                            
 redistribute_writes                          | true                  | true                  | boolean | Force parallel distributed writes                                                                                      
 reorder_joins                                |                       |                       | boolean | (DEPRECATED) Reorder joins to remove unnecessary cross joins. If this is set, join_reordering_strategy will be ignored 
 resource_overcommit                          | false                 | false                 | boolean | Use resources which are not guaranteed to be available to the query                                                    
 scale_writers                                | false                 | false                 | boolean | Scale out writers based on throughput (use minimum necessary)                                                          
 spatial_join                                 | true                  | true                  | boolean | Use spatial index for spatial join when possible                                                                       
 spatial_partitioning_table_name              |                       |                       | varchar | Name of the table containing spatial partitioning scheme                                                               
 spill_enabled                                | false                 | false                 | boolean | Experimental: Enable spilling                                                                                          
 split_concurrency_adjustment_interval        | 100.00ms              | 100.00ms              | varchar | Experimental: Interval between changes to the number of concurrent splits per node                                     
 statistics_cpu_timer_enabled                 | true                  | true                  | boolean | Experimental: Enable cpu time tracking for automatic column statistics collection on write                             
 task_concurrency                             | 4                     | 4                     | bigint  | Default number of local parallel jobs per worker                                                                       
 task_share_index_loading                     | false                 | false                 | boolean | Share index join lookups and caching within a task                                                                     
 task_writer_count                            | 1                     | 1                     | bigint  | Default number of local parallel table writer jobs per worker                                                          
 test_long                                    | 42                    | 42                    | bigint  | test long property                                                                                                     
 test_string                                  | test default          | test default          | varchar | test string property                                                                                                   
 use_mark_distinct                            | true                  | true                  | boolean | Implement DISTINCT aggregations using MarkDistinct                                                                     
 writer_min_size                              | 32MB                  | 32MB                  | varchar | Target minimum size of writer output when scaling writers                                                              
 testing_catalog.connector_boolean            | true                  | true                  | boolean | connector boolean property                                                                                             
 testing_catalog.connector_double             | 99.0                  | 99.0                  | double  | connector double property                                                                                              
 testing_catalog.connector_long               | 33                    | 33                    | bigint  | connector long property                                                                                                
 testing_catalog.connector_string             | connector default     | connector default     | varchar | connector string property                                                                                              
(67 rows)

Query 20190130_032945_00000_jd8pg, FINISHED, 1 node
http://localhost:8080/ui/query.html?20190130_032945_00000_jd8pg
Splits: 5 total, 5 done (100.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 46% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/sParallelism: 0.0
Peak Memory: 0B0:03 [0 rows, 0B] [0 rows/s, 0B/s]
```

Reply:
@mbasmanova Masha, I was just using HiveQueryRunner and the built CLI on the aria-deb branch. For queries like show session, the FilterNode was firstly created with the predicateWithoutTupleDomain null (the constructor that does not take the predicateWithoutTupleDomain parameter was called). This happens in QueryPlanner.filter(). Then FilterNode.replaceChildren() was called repeatedly on the previously created FilterNode, which has predicateWithoutTupleDomain null. So the predicateWithoutTupleDomain field is nullable. For such case, I don't think give it a non null value (e.g. true) is  semantically correct. What do you think?

Reply:
![image](https://user-images.githubusercontent.com/33299678/51959465-6da58c80-248f-11e9-8f6e-6220aab85bc9.png)
This is the call stack for the first time the FilterNode was created with a null predicateWithoutTupleDomain

Reply:
@yingsu00 Ying, looks like you are working off of stale version of aria-dev branch. In the current version,` FilterNode` constructor initializes `predicateWithoutTupleDomain` to `TRUE_LITERAL`.

```
    public FilterNode(PlanNodeId id,
                       PlanNode source,
                       Expression predicate)
    {
        super(id);

        this.source = source;
        this.predicate = predicate;
        this.predicateWithoutTupleDomain = TRUE_LITERAL;
    }
```

Here is a commit that made that change:

https://github.com/prestodb/presto/commit/b44fc183eb84ba39801897baa64c6071d2893eee#diff-094bef3900c7781b9b212ed1d8041cce